File: HackerRank/30 Days of Code/Day19_Interfaces.java
// Day 19. Interfaces
// https://www.hackerrank.com/challenges/30-interfaces/problem

import java.io.*;
import java.util.*;

interface AdvancedArithmetic{
   int divisorSum(int n);
}
class Calculator implements AdvancedArithmetic {
    @Override
    public int divisorSum(int n) {
        int res = 0;
        for (int i = 1; i <= n; i++)
            if (n % i == 0)
                res += i;
        return res;
    }
}

class Solution {

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        int n = scan.nextInt();
        scan.close();
        
      	AdvancedArithmetic myCalculator = new Calculator(); 
        int sum = myCalculator.divisorSum(n);
        System.out.println("I implemented: " + myCalculator.getClass().getInterfaces()[0].getName() );
        System.out.println(sum);
    }
}

// 인터페이스 상속하여 실체화하기
// interface로 인터페이스 선언, 이를 implements하는 클래스는 인터페이스의 메서드를 모두 @Override해야 함

File: HackerRank/30 Days of Code/Day19_Interfaces.py
# Day 19. Interfaces
# https://www.hackerrank.com/challenges/30-interfaces/problem

class AdvancedArithmetic(object):
    def divisorSum(n):
        raise NotImplementedError

class Calculator(AdvancedArithmetic):
    def divisorSum(self, n):
        res = 0
        for i in range(1, n + 1):
            if n % i == 0:
                res += i
        return res


n = int(input())
my_calculator = Calculator()
s = my_calculator.divisorSum(n)
print("I implemented: " + type(my_calculator).__bases__[0].__name__)
print(s)

# 인터페이스 상속하여 실체화하기
# python에서는 별도의 인터페이스 클래스가 없고, 대신 인터페이스 클래스에서 오버라이딩하지 않으면 못 쓰도록 오류 걸어놓기